### PR TITLE
Create master/compute substacks and remove CloudFormation Init when bootstrapping compute fleet

### PR DESCRIPTION
**Pull request by demartinofra**
## Changes

Here is a summary of the changes in this PR, commit by commit:

**Extract master server and compute fleet in separate CFN substacks**

This only extracts the head node and compute fleet related resources to separate CloudFormation nested stacks. The reason behind this is that it simplifies the changes applied in the following commits. No functional changes are expected as part of this commit.
Some of the resources that are specific to compute nodes or head node (such as subnets) are still defined in the main stack in order to avoid that other dependent resources (e.g EFSSubstack) have to wait longer than actually needed.

***

**Explicitly pass IAM role when calling cfn-init/signal commands**

If no credentials are specified, AWS CloudFormation checks for stack membership and limits the scope of the call to the stack that the instance belongs to.
Explicitly passing the instance role name to `cfn-init` and `cfn-signal` calls.

When this gets released we need to update the ParallelClusterInstancePolicy in the docs by adding "cloudformation:SignalResource".

***

**Update uploadTemplate script to include master and compute-fleet substacks** 

Simple update to the uploadTemplate script to account for the new substacks.

***

**Add ASG name to the stacks output**

Adding ASG to stack output so that it can be retrieved from there rather than parsing through stack resources.

***

**Convert compute-fleet-substack and master-server-substack to YAML**

Used cfn-flip and the following commands for the conversion:
- cfn-flip -y -c master-server-substack.cfn.json master-server-substack.cfn.yaml
- cfn-flip -y -c compute-fleet-substack.cfn.json compute-fleet-substack.cfn.yaml

***

**Add formatting check to tox/travis for YAML CloudFormation docs**

Applied formatting checks to the new YAML files.

***

**Add aws-parallelcluster-node-type tag to cluster instances**

Name tag is more likely to be changed by the user, hence adding a new tag to identify the node type.

***

**Make pcluster CLI commands compatible with the master/compute substacks**

- Use EC2 API rather than CloudFormation and ASG in order to retrieve cluster nodes (master and compute). The new logic filters nodes based on the new `aws-parallelcluster-node-type` tag and has fallback behaviors to preserve backward compatibility. 

***

**integ-tests: minor changes to make them compatible with new templates**

Update logic that retrieves ASG name and master instance ip.

***

**compute-fleet: move generation of bootstrap files to userdata**

Move the generation of files needed in the bootstrap phase from the cfninit metadata to userdata

***

**Remove CloudFormation Init from compute fleet userdata**

Replace cfninit logic with plain userdata in ComputeFleet

***

**Append user-data output to the EC2 console log**

Add userdata logging to the console output so that in case the user is unable to ssh into the instance some trouble-shooting can be performed from the EC2 console.

Here is an extract from console system logs:
```
<13>Jun 15 09:47:24 user-data: Running handlers:
<13>Jun 15 09:47:24 user-data: Running handlers complete
<13>Jun 15 09:47:24 user-data: Chef Client finished, 3/3 resources updated in 05 seconds
<13>Jun 15 09:47:24 user-data: + popd
<13>Jun 15 09:47:24 user-data: /tmp
<13>Jun 15 09:47:24 user-data: + '[' '!' -f /opt/parallelcluster/.bootstrapped ']'
<13>Jun 15 09:47:24 user-data: + cfn-signal --exit-code=0 '--reason=ComputeServer setup complete' --stack=*** --role=*** --resource=ComputeFleet --region=us-east-2

```

***

**Remove polling of WaitCondition from compute nodes**

This was introduced as an optimization to let the compute fleet start before the master had totally completed its bootstrap.
The optimization was relying on the CloudFormation API that has a very low TPS, hence it was limiting the scaling speed when hundreds of nodes are launched at once.
Dropping it for now. If we want to re-introduce this the status of the master node needs to be fetched from a more scalable source.

***

**Increase MasterServer and ComputeFleet timeout to 60 minutes**

Increase the provisioning timeout to account for long running user-data. 

***

## Testing

Executed the following integ tests: `cfn-init cloudwatch_logging dcv disable_hyperthreading efa slurm sge torque awsbatch spot update ebs efs fsx_lustre raid`

***
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cloudformation/compute-fleet-substack.cfn.yaml [hunk 1]
```diff
@@ -440,8 +440,21 @@ Resources:
...
+                pushd /etc/chef &&
+                chef-client --local-mode --config /etc/chef/client.rb --log_level auto --force-formatter --no-color --chef-zero-port 8889 --json-attributes /etc/chef/dna.json --override-runlist aws-parallelcluster::_prep_env &&
+                /opt/parallelcluster/scripts/fetch_and_run -preinstall &&
+                chef-client --local-mode --config /etc/chef/client.rb --log_level auto --force-formatter --no-color --chef-zero-port 8889 --json-attributes /etc/chef/dna.json &&
+                /opt/parallelcluster/scripts/fetch_and_run -postinstall &&
+                chef-client --local-mode --config /etc/chef/client.rb --log_level auto --force-formatter --no-color --chef-zero-port 8889 --json-attributes /etc/chef/dna.json --override-runlist aws-parallelcluster::finalize &&
+                popd
+              } || error_exit 'Failed to run bootstrap recipes. If --norollback was specified, check /var/log/cfn-init.log and /var/log/cloud-init-output.log.'
+
+              [ ! -f /opt/parallelcluster/.bootstrapped ] && echo ${!parallelcluster_version} | tee /opt/parallelcluster/.bootstrapped
```
Comment (lukeseawalker) [inline cloudformation/compute-fleet-substack.cfn.yaml:452]: is this `${!parallelcluster_version}` wanted?

**Comment by demartinofra** [comment]
> Does the upload-cfn-template.py utility need to be updated as well?

yes! done

File: util/upload-cfn-templates.py [hunk 2]
```diff
@@ -23,6 +25,19 @@ def get_all_aws_regions(region):
     return set(sorted(r.get("RegionName") for r in ec2.describe_regions().get("Regions")))
 
 
+def get_template_extension(templates_dir, template_name):
+    matching_files = glob("{}{}.*".format(templates_dir, template_name))
+    if len(matching_files) != 1:
```
Comment (ddeidda) [inline util/upload-cfn-templates.py:30]: Is this to exclude that, by chance, both the .json and the .yaml template exist for the same template_name?
minor: I am wondering, why doesn't this script simply use a map to recognize the only valid template files expected for substacks?

File: cli/tests/pcluster/utils/test_pcluster_utils.py [hunk 3]
```diff
@@ -388,3 +388,85 @@ def test_create_s3_bucket(region, error_message, boto3_stubber):
             utils.create_s3_bucket(bucket_name, region)
     else:
         utils.create_s3_bucket(bucket_name, region)
+
+
+@pytest.mark.parametrize(
+    "node_type, expected_fallback, expected_response, expected_instances",
+    [
+        (utils.NodeType.master, False, {"Reservations": [{"Groups": [], "Instances": [{}]}]}, 1),
```
[the remark this answers is not in the document]
Comment (demartinofra) [reply cli/tests/pcluster/utils/test_pcluster_utils.py:396]: The empty objects `{}` inside the instances array are considered valid instances for the test hence in those cases the fallback is not used (see also the `expected_fallback` parameter).

Here you can double check the test coverage: https://codecov.io/gh/aws/aws-parallelcluster/pull/1826/src/cli/pcluster/utils.py?before=cli/pcluster/utils.py

File: cloudformation/master-server-substack.cfn.yaml [hunk 4]
```diff
@@ -181,6 +181,10 @@ Resources:
       LaunchTemplate:
         LaunchTemplateId: !Ref 'MasterServerLaunchTemplate'
         Version: !GetAtt 'MasterServerLaunchTemplate.LatestVersionNumber'
+    CreationPolicy:
```
[the remark this answers is not in the document]
Comment (demartinofra) [reply cloudformation/master-server-substack.cfn.yaml:184]: yes it does

**Pull request merged**